Sources/_StringProcessing/Engine/MEQuantify.swift

```diff
@@ -0,0 +1,122 @@
+  /// - Handles .eager and .posessive
+  /// - Handles arbitrary minTrips and extraTrips
+  mutating func runQuantify(_ payload: QuantifyPayload) -> Bool {
+    var trips = 0
+    var extraTrips = payload.extraTrips
+    var savePoint = startQuantifierSavePoint()
+    
+    while true {
+      if trips >= payload.minTrips {
+        if extraTrips == 0 { break }
+        extraTrips = extraTrips.map({$0 - 1})
+        if payload.quantKind == .eager {
+          savePoint.updateRange(newEnd: currentPosition)
+        }
+      }
+      let next = _doQuantifyMatch(payload)
+      guard let idx = next else { break }
+      currentPosition = idx
+      trips += 1
+    }
+    
+    if trips < payload.minTrips {
+      signalFailure()
+      return false
+    }
+    
+    if payload.quantKind == .eager && !savePoint.rangeIsEmpty {
+      // The last save point has saved the current position, so it's unneeded
+      savePoint.shrinkRange(input)
+      if !savePoint.rangeIsEmpty {
+        savePoints.append(savePoint)
+      }
+    }
+    return true
+  }
+  
+  /// Specialized quantify instruction interpreter for *
+  mutating func runEagerZeroOrMoreQuantify(_ payload: QuantifyPayload) -> Bool {
+    assert(payload.quantKind == .eager
+           && payload.minTrips == 0
+           && payload.extraTrips == nil)
+    var savePoint = startQuantifierSavePoint()
+    
+    while true {
+      savePoint.updateRange(newEnd: currentPosition)
+      let next = _doQuantifyMatch(payload)
+      guard let idx = next else { break }
+      currentPosition = idx
+    }
+    
+    // The last save point has saved the current position, so it's unneeded
+    savePoint.shrinkRange(input)
+    if !savePoint.rangeIsEmpty {
+      savePoints.append(savePoint)
+    }
+    return true
+  }
+  
+  /// Specialized quantify instruction interpreter for +
+  mutating func runEagerOneOrMoreQuantify(_ payload: QuantifyPayload) -> Bool {
+    assert(payload.quantKind == .eager
+           && payload.minTrips == 1
+           && payload.extraTrips == nil)
+    var savePoint = startQuantifierSavePoint()
+    while true {
+      let next = _doQuantifyMatch(payload)
+      guard let idx = next else { break }
+      currentPosition = idx
+      savePoint.updateRange(newEnd: currentPosition)
+    }
+    
+    if savePoint.rangeIsEmpty {
+      signalFailure()
+      return false
+    }
+    // The last save point has saved the current position, so it's unneeded
+    savePoint.shrinkRange(input)
+    if !savePoint.rangeIsEmpty {
+      savePoints.append(savePoint)
+    }
+    return true
+  }
+  
+  /// Specialized quantify instruction interpreter for ?
+  mutating func runZeroOrOneQuantify(_ payload: QuantifyPayload) -> Bool {
+    assert(payload.minTrips == 0
+           && payload.extraTrips == 1)
+    let next = _doQuantifyMatch(payload)
+    guard let idx = next else {
+      return true // matched zero times
+    }
+    if payload.quantKind != .possessive {
```

Comment:
Do these work for reluctant? Possessive also never backtracks, so I wonder if (future work) we should consider it entirely separately.

Reply:
Reluctant quantification isn't allowed into `.quantify` because it never loops inside it like the other two so it didn't make sense to add into `.quantify`

In the future we could have a specialized reluctant quantifier instruction that takes in a both the quantification and the match instruction after it, which would handle cases like `.*?;`

Reply:
Can you add some asserts then?

Reply:
Oops, I had some before but they got lost in the other changes

Reply:
Future work: It should be possible to optimize the pattern of reluctant quantifier + anchor character in a similar style by having a payload of two int registers that store the respective payloads for the quantification and the anchor.

The reason why we can't do this peephole optimization now is that we're stuck with the tree representation so there isn't a good way of determining if the reluctant quantifier is followed by an anchor